Inconsistent results between runs

I've been working on a project which takes the results from periphery and adds comments to the correct lines of code so that my team are aware of the issues that are found and we can correct these issues if we are working in the same area. However, I've run into a problem. If I run periphery, add comments, then run it again, I get different results the second time round. 

I control for lines moving around etc. I filter out all results in files that are in multiple targets. I ignore `@objc` annotations. 

I'm finding that even in the raw results, on the first run it can find `N` results in a given file, then `M` results on a second run when `M` != `N`. There's no consistency in it (i.e. it's not always protocols or method parameters, etc.) as far as I can tell, and I've been unable to find a minimal repro. I also can't share our project unfortunately. I can tell you that it is ~1,000,000 lines of code though. If there's any way I can start to narrow down what's going on, I'd love to know so I can provide that info here. 

Sorry that this isn't a nice concrete issue with an easy repro! 

Edit: To clarify, it's <2% of the results that are different.

I see (probably) the same inconsistencies between runs even for this simple example:
```
protocol FooProtocol {
    var a: Int? { get set }
    func doIt()
}

final class Foo: FooProtocol {
    var a: Int?
    func doIt() {
        print(a!)
    }
}

class Bar {
    private var foo: FooProtocol
    
    init(foo: FooProtocol = Foo()) {
        self.foo = foo
    }
    
    func doIt() {
        foo.a = 3
        foo.doIt()
    }
}

let bar = Bar()
bar.doIt()
```
gives me in about 10%-40% of the runs

```
main.swift:2:9: warning: Property 'a' is assigned, but never used
```
This file is a member of the target I pass to --targets
swift version is 5.3.2
I'm using periphery 2.4.3 (latest)

This is the output of 10 consecutive runs:

```
~/tmp/periphery periphery scan
* Inspecting project...
* Building Periphery...
* Indexing...
* Analyzing...

* No unused code detected.
~/tmp/periphery periphery scan
* Inspecting project...
* Building Periphery...
* Indexing...
* Analyzing...

* No unused code detected.
~/tmp/periphery periphery scan
* Inspecting project...
* Building Periphery...
* Indexing...
* Analyzing...

* No unused code detected.
~/tmp/periphery periphery scan
* Inspecting project...
* Building Periphery...
* Indexing...
* Analyzing...

main.swift:2:9: warning: Property 'a' is assigned, but never used

* Seeing false positives?
 - Periphery only analyzes files that are members of the targets you specify.
   References to declarations identified as unused may reside in files that are members of other targets, e.g test targets.
 - By default, Periphery does not assume that all public declarations are in use. 
   You can instruct it to do so with the --retain-public option.
 - Periphery is a very precise tool, false positives often turn out to be correct after further investigation.
 - If it really is a false positive, please report it - https://github.com/peripheryapp/periphery/issues.
~/tmp/periphery periphery scan
* Inspecting project...
* Building Periphery...
* Indexing...
* Analyzing...

main.swift:2:9: warning: Property 'a' is assigned, but never used

* Seeing false positives?
 - Periphery only analyzes files that are members of the targets you specify.
   References to declarations identified as unused may reside in files that are members of other targets, e.g test targets.
 - By default, Periphery does not assume that all public declarations are in use. 
   You can instruct it to do so with the --retain-public option.
 - Periphery is a very precise tool, false positives often turn out to be correct after further investigation.
 - If it really is a false positive, please report it - https://github.com/peripheryapp/periphery/issues.
~/tmp/periphery periphery scan
* Inspecting project...
* Building Periphery...
* Indexing...
* Analyzing...

main.swift:2:9: warning: Property 'a' is assigned, but never used

* Seeing false positives?
 - Periphery only analyzes files that are members of the targets you specify.
   References to declarations identified as unused may reside in files that are members of other targets, e.g test targets.
 - By default, Periphery does not assume that all public declarations are in use. 
   You can instruct it to do so with the --retain-public option.
 - Periphery is a very precise tool, false positives often turn out to be correct after further investigation.
 - If it really is a false positive, please report it - https://github.com/peripheryapp/periphery/issues.
~/tmp/periphery periphery scan
* Inspecting project...
* Building Periphery...
* Indexing...
* Analyzing...

* No unused code detected.
~/tmp/periphery periphery scan
* Inspecting project...
* Building Periphery...
* Indexing...
* Analyzing...

main.swift:2:9: warning: Property 'a' is assigned, but never used

* Seeing false positives?
 - Periphery only analyzes files that are members of the targets you specify.
   References to declarations identified as unused may reside in files that are members of other targets, e.g test targets.
 - By default, Periphery does not assume that all public declarations are in use. 
   You can instruct it to do so with the --retain-public option.
 - Periphery is a very precise tool, false positives often turn out to be correct after further investigation.
 - If it really is a false positive, please report it - https://github.com/peripheryapp/periphery/issues.
~/tmp/periphery periphery scan
* Inspecting project...
* Building Periphery...
* Indexing...
* Analyzing...

* No unused code detected.
~/tmp/periphery periphery scan
* Inspecting project...
* Building Periphery...
* Indexing...
* Analyzing...

* No unused code detected.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,7 +11,7 @@
 
 ##### Bug Fixes
 
-- None.
+- Fix issue where protocol property references could be incorrectly associated with the getter/setter rather than the property itself, leading to erroneous results.
 
 ## 2.4.3 (2021-03-05)
 

--- a/Sources/PeripheryKit/Analyzer/Visitors/AssignOnlyPropertyReferenceEliminator.swift
+++ b/Sources/PeripheryKit/Analyzer/Visitors/AssignOnlyPropertyReferenceEliminator.swift
@@ -19,7 +19,7 @@ final class AssignOnlyPropertyReferenceEliminator: SourceGraphVisitor {
 
         let setters = graph.declarations(ofKind: .functionAccessorSetter)
 
-        var assignOnlyProperties: [(Declaration, Reference)] = []
+        var assignOnlyProperties: [Declaration: [Reference]] = [:]
 
         for setter in setters {
             let references = graph.references(to: setter)
@@ -38,15 +38,22 @@ final class AssignOnlyPropertyReferenceEliminator: SourceGraphVisitor {
                 let hasGetterReference = caller.references.contains { $0.kind == .functionAccessorGetter && $0.name == getterName }
 
                 if !hasGetterReference {
-                    assignOnlyProperties.append((property, propertyReference))
+                    assignOnlyProperties[property, default: []].append(propertyReference)
                 }
             }
         }
 
         var retainedProperties: Set<Declaration> = []
+        let allProperties = Set(assignOnlyProperties.map { $0.0 })
+
+        // A protocol property can technically be assigned and never used when the protocol is used as an existential
+        // type, however communicating that succinctly would be very tricky, and most likely just lead to confusion.
+        // Here we filter out protocol properties and thus restrict this analysis only to concrete properties.
+        allProperties
+            .filter { ($0.parent as? Declaration)?.kind == .protocol }
+            .forEach { retainedProperties.insert($0) }
 
         if !configuration.retainAssignOnlyPropertyTypes.isEmpty {
-            let allProperties = Set(assignOnlyProperties.map { $0.0 })
             let propertiesByType = try PropertyTypeParser.parse(allProperties)
 
             for (type, properties) in propertiesByType {
@@ -56,10 +63,13 @@ final class AssignOnlyPropertyReferenceEliminator: SourceGraphVisitor {
             }
         }
 
-        for (property, reference) in assignOnlyProperties {
+        for (property, references) in assignOnlyProperties {
             if !retainedProperties.contains(property) {
-                graph.remove(reference)
                 property.analyzerHint = .assignOnlyProperty
+
+                for reference in references {
+                    graph.remove(reference)
+                }
             }
         }
     }

--- a/Sources/PeripheryKit/Indexer/SourceGraph.swift
+++ b/Sources/PeripheryKit/Indexer/SourceGraph.swift
@@ -154,6 +154,11 @@ public final class SourceGraph {
         }
     }
 
+    func makeImplicitUnsafe(_ declaration: Declaration) {
+        declaration.isImplicit = true
+        declaration.usrs.forEach { allExplicitDeclarationsByUsr.removeValue(forKey: $0) }
+    }
+
     func markReachable(_ declaration: Declaration) {
         mutationQueue.sync {
             _ = reachableDeclarations.insert(declaration)

--- a/Sources/PeripheryKit/Indexer/SwiftIndexer.swift
+++ b/Sources/PeripheryKit/Indexer/SwiftIndexer.swift
@@ -200,6 +200,7 @@ public final class SwiftIndexer {
             }
 
             establishDeclarationHierarchy()
+            makeProtocolPropertyAccessorsImplicit(with: decls)
             associateDanglingReferences(with: decls)
 
             let syntax = try SyntaxParser.parse(file.url)
@@ -262,6 +263,22 @@ public final class SwiftIndexer {
                         }
                     }
                 }
+            }
+        }
+
+        // All property accessors with a body are considered explicit, because they are explicitly declared in code.
+        // Property accessors without a body are still present, however they are implied by the compiler and thus
+        // marked as implicit. Protocol property accessors muddy this distinction, as they are explicitly declared
+        // e.g: `var foo: Int { get set }` yet they cannot have a body. Swift marks these accessors as explicit,
+        // however for our intents and purposes we need to consider them as implicit.
+        private func makeProtocolPropertyAccessorsImplicit(with decls: [Declaration]) {
+            graph.mutating {
+                decls.filter { $0.kind == .protocol }
+                    .flatMap { $0.declarations }
+                    .filter { $0.kind.isVariableKind }
+                    .flatMap { $0.declarations }
+                    .filter { $0.kind.isAccessorKind }
+                    .forEach { graph.makeImplicitUnsafe($0) }
             }
         }
 

--- a/Tests/PeripheryTests/RetentionTest.swift
+++ b/Tests/PeripheryTests/RetentionTest.swift
@@ -525,6 +525,13 @@ class RetentionTest: SourceGraphTestCase {
         }
     }
 
+    func testAssignOnlyPropertyAnalysisDoesNotApplyToProtocolProperties() {
+        analyze(retainPublic: true) {
+            XCTAssertReferenced((.varInstance, "someProperty"), descendentOf: (.protocol, "FixtureProtocol124"))
+            XCTAssertReferenced((.varInstance, "someProperty"), descendentOf: (.class, "FixtureClass124"))
+        }
+    }
+
     func testPropertyReferencedByComputedValue() {
         analyze(retainPublic: true) {
             XCTAssertReferenced((.class, "FixtureClass78"))

--- a/Tests/RetentionFixtures/testAssignOnlyPropertyAnalysisDoesNotApplyToProtocolProperties.swift
+++ b/Tests/RetentionFixtures/testAssignOnlyPropertyAnalysisDoesNotApplyToProtocolProperties.swift
@@ -1,0 +1,16 @@
+import Foundation
+
+protocol FixtureProtocol124 {
+    var someProperty: Int? { get set }
+}
+
+class FixtureClass124: FixtureProtocol124 {
+    var someProperty: Int?
+}
+
+public class FixtureClass124Retainer {
+    public func retain() {
+        var cls: FixtureProtocol124 = FixtureClass124()
+        cls.someProperty = 1
+    }
+}

